core/src/main/java/cn/edu/tsinghua/iot/benchmark/workload/SingletonWorkDataWorkLoad.java

```diff
@@ -62,29 +64,41 @@ public static SingletonWorkDataWorkLoad getInstance() {
   }
 
   @Override
-  public Batch getOneBatch() throws WorkloadException {
-    long curLoop = insertLoop.getAndIncrement();
-    Batch batch = new Batch();
-    // create schema of batch
-    List<Sensor> sensors = new ArrayList<>();
-    if (config.isIS_SENSOR_TS_ALIGNMENT()) {
-      sensors = SENSORS;
+  public IBatch getOneBatch() throws WorkloadException {
+    IBatch batch = null;
+    final int recordsNumPerDevice = config.getBATCH_SIZE_PER_WRITE() / config.getDEVICE_NUM_PER_WRITE();
+    if (config.getDEVICE_NUM_PER_WRITE() == 1) {
+      batch = new Batch();
     } else {
-      int sensorId = sensorIndex.getAndIncrement() % config.getSENSOR_NUMBER();
-      batch.setColIndex(sensorId);
-      sensors.add(SENSORS.get(sensorId));
+      batch = new MultiDeviceBatch(config.getDEVICE_NUM_PER_WRITE());
     }
-    DeviceSchema deviceSchema =
-        new DeviceSchema(
-            MetaUtil.getDeviceId((int) curLoop % config.getDEVICE_NUMBER()),
-            sensors,
-            config.getDEVICE_TAGS());
-    batch.setDeviceSchema(deviceSchema);
-    // create data of batch
-    for (long batchOffset = 0; batchOffset < config.getBATCH_SIZE_PER_WRITE(); batchOffset++) {
-      long stepOffset =
-          (curLoop / config.getDEVICE_NUMBER()) * config.getBATCH_SIZE_PER_WRITE() + batchOffset;
-      addOneRowIntoBatch(batch, stepOffset);
+    for (int deviceID = 0; deviceID < config.getDEVICE_NUM_PER_WRITE(); deviceID++) {
+      long curLoop = insertLoop.getAndIncrement();
+      // create schema of batch
+      List<Sensor> sensors = new ArrayList<>();
+      if (config.isIS_SENSOR_TS_ALIGNMENT()) {
+        sensors = SENSORS;
+      } else {
+        int sensorId = sensorIndex.getAndIncrement() % config.getSENSOR_NUMBER();
+        batch.setColIndex(sensorId);
+        sensors.add(SENSORS.get(sensorId));
+      }
+      DeviceSchema deviceSchema =
+          new DeviceSchema(
+              MetaUtil.getDeviceId((int) curLoop % config.getDEVICE_NUMBER()),
+              sensors,
+              config.getDEVICE_TAGS());
+      batch.setDeviceSchema(deviceSchema);
+      // create data of batch
+      for (long batchOffset = 0; batchOffset < recordsNumPerDevice; batchOffset++) {
+        long stepOffset =
+            (curLoop / config.getDEVICE_NUMBER()) * config.getBATCH_SIZE_PER_WRITE()
+                + batchOffset;
+        addOneRowIntoBatch(batch, stepOffset);
+      }
+      if (batch.hasNext()) {
+        batch.next();
+      }
```

Comment:
A little bit strange here, I think we need to consider setDeviceSchema method in IBatch

Reply:
We should discuss this part later